### DISCUSSION TITLE
signal pattern

### DISCUSSION THREAD

**Discussion by mattkasun** [mochi-mqtt/server#268] [General]
All of the examples use the following pattern:
```
func main() {
	sigs := make(chan os.Signal, 1)
	done := make(chan bool, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigs
		done <- true
	}()
// rest of main
        <-done
	server.Log.Warn().Msg("caught signal, stopping...")
	server.Close()
	server.Log.Info().Msg("main.go finished")
}
```

Is there an advantage to this pattern vs
```
func main() {
	done := make(chan os.Signal, 1)
	signal.Notify(done, syscall.SIGINT, syscall.SIGTERM)
// rest of main
       <-done
       server.Log.Warn().Msg("caught signal, stopping...")	
       server.Close()
       server.Log.Info().Msg("main.go finished")
}
```

**Comment by dgduncan** [comment]
@mattkasun I think as it stands right now there is no advantage or disadvantage, minus the extremely slight overhead of having an additional go routine and channel. Others can chime in if they know better, but I am willing to bet that this is a placeholder for if in the future this done channel wants to be used by anything else. 

i.e if in the future we want the ability to pass that `done` into a struct such that in event of a critical error the broker can shut down gracefully passing a `done` channel makes more sense than passing a `signal` channel.

But as it stands I believe you are correct in that the functionality can be done just by using the `signal` compared to the `done` and `signal`.
